### PR TITLE
[SPARK-12310] [SparkR] Add write.json and write.parquet for SparkR

### PR DESCRIPTION
**Pull request by yanboliang**
Add `write.json` and `write.parquet` for SparkR, and deprecated `saveAsParquetFile`.

### CODE REVIEW

File: R/pkg/R/DataFrame.R [hunk 1]
```diff
@@ -614,12 +641,25 @@ setMethod("toJSON",
...
+#' saveAsParquetFile(df, "/tmp/sparkr-tmp2/")
 #'}
+setMethod("write.parquet",
+          signature(x = "DataFrame", path = "character"),
+          function(x, path) {
+            write <- callJMethod(x@sdf, "write")
+            invisible(callJMethod(write, "parquet", path))
+          })
+
+#' @family DataFrame functions
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: putting @family here would create multiple "See Also" section in the resulting html page, we would only need one for all those with the same @rdname

File: R/pkg/inst/tests/testthat/test_sparkSQL.R [hunk 2]
```diff
@@ -371,13 +371,13 @@ test_that("Collect DataFrame with complex types", {
...
 
-test_that("read.json()/jsonFile() on a local file returns a DataFrame", {
+test_that("read.json()/jsonFile() and write.json()", {
   df <- read.json(sqlContext, jsonPath)
   expect_is(df, "DataFrame")
   expect_equal(count(df), 3)
   # read.json()/jsonFile() works with multiple input paths
   jsonPath2 <- tempfile(pattern="jsonPath2", fileext=".json")
-  write.df(df, jsonPath2, "json", mode="overwrite")
+  write.json(df, jsonPath2)
```
Comment (felixcheung) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: I think it actually make sense to have a test for `write.df` with additional option (in this case `mode="overwrite"`) - is there a reason for removing this?

File: R/pkg/inst/tests/testthat/test_sparkSQL.R [hunk 3]
```diff
@@ -454,6 +454,8 @@ test_that("insertInto() on a registered table", {
   expect_equal(count(sql(sqlContext, "select * from table1")), 2)
   expect_equal(first(sql(sqlContext, "select * from table1 order by age"))$name, "Bob")
   dropTempTable(sqlContext, "table1")
+
+  unlink(parquetPath2)
```
Comment (felixcheung) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: should there be `unlink(jsonPath2)` as well? if so could you please add that

File: R/pkg/inst/tests/testthat/test_sparkSQL.R [hunk 4]
```diff
@@ -371,13 +371,13 @@ test_that("Collect DataFrame with complex types", {
   expect_equal(bob$height, 176.5)
 })
 
-test_that("read.json()/jsonFile() on a local file returns a DataFrame", {
```
Comment (sun-rui) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: How about renaming this test case to "read/write json files", where read.json(), jsonFile(), read.df(), write.json(), write.df() are tested to read/write json files. This test case can be merged with the one "read.df() from json file"

File: R/pkg/inst/tests/testthat/test_sparkSQL.R [hunk 5]
```diff
@@ -1448,23 +1457,28 @@ test_that("write.df() on DataFrame and works with read.parquet", {
   expect_equal(count(df), count(parquetDF))
 })
 
-test_that("read.parquet()/parquetFile() works with multiple input paths", {
+test_that("read.parquet()/parquetFile() and write.parquet()/saveAsParquetFile()", {
```
Comment (sun-rui) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: How about renaming this test case to "read/write Parquet files", where read.parquet(), parquetFile(), read.df(), write.parquet(), saveAsParquetFile(), write.df() are tested to read/write json files. This test case can be merged with the one "write.df() as parquet file"

File: R/pkg/inst/tests/testthat/test_sparkSQL.R (hunk 2 above)
Comment (yanboliang) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: I agree to have a test for ```write.df```, but it still exists even I removing this such as [here](https://github.com/apache/spark/blob/master/R/pkg/inst/tests/testthat/test_sparkSQL.R#L870).

File: R/pkg/R/DataFrame.R (hunk 1 above)
Comment (yanboliang) [inline R/pkg/R/DataFrame.R]: Good point.

File: R/pkg/inst/tests/testthat/test_sparkSQL.R (hunk 5 above)
Comment (yanboliang) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: I will reorg the test cases, thanks!

**Comment by sun-rui** [comment]
LGTM

**Comment by shivaram** [comment]
LGTM. Thanks @yanboliang and @sun-rui -- Merging this to master and branch-1.6

**Pull request closed without merging**